base/abstractarray.jl

```diff
@@ -144,77 +150,82 @@ linearindexing(::LinearFast, ::LinearFast) = LinearFast()
+# specialize on array type, and try to avoid specializations on index
+# types; conversely, `checkindex` is intended to be specialized only
+# on index type (especially, its last argument).
+
 """
-    checkindex(Bool, inds::UnitRange, index)
+    checkbounds(Bool, A, I...)
 
-Return `true` if the given `index` is within the bounds of
-`inds`. Custom types that would like to behave as indices for all
-arrays can extend this method in order to provide a specialized bounds
-checking implementation.
+Return `true` if the specified indices `I` are in bounds for the given
+array `A`. Subtypes of `AbstractArray` should specialize this method
+if they need to provide custom bounds checking behaviors; however, in
+many cases one can rely on `A`'s indices and `checkindex`.
+
+See also `checkindex`.
 """
-checkindex(::Type{Bool}, inds::AbstractUnitRange, i) = throw(ArgumentError("unable to check bounds for indices of type $(typeof(i))"))
-checkindex(::Type{Bool}, inds::AbstractUnitRange, i::Real) = (first(inds) <= i) & (i <= last(inds))
-checkindex(::Type{Bool}, inds::AbstractUnitRange, ::Colon) = true
-function checkindex(::Type{Bool}, inds::AbstractUnitRange, r::Range)
-    @_propagate_inbounds_meta
-    isempty(r) | (checkindex(Bool, inds, first(r)) & checkindex(Bool, inds, last(r)))
+function checkbounds(::Type{Bool}, A::AbstractArray, I...)
+    @_inline_meta
+    checkbounds_indices(indices(A), I)
 end
-checkindex{N}(::Type{Bool}, indx::AbstractUnitRange, I::AbstractArray{Bool,N}) = N == 1 && indx == indices1(I)
-function checkindex(::Type{Bool}, inds::AbstractUnitRange, I::AbstractArray)
+function checkbounds(::Type{Bool}, A::AbstractArray, I::AbstractArray{Bool})
     @_inline_meta
-    b = true
-    for i in I
-        b &= checkindex(Bool, inds, i)
-    end
-    b
+    checkbounds_logical(A, I)
 end
 
-# check all indices/dimensions
-# To make extension easier, avoid specializations of checkbounds on index types
-# (That said, typically one does not need to specialize this function.)
 """
-    checkbounds(Bool, array, indexes...)
+    checkbounds_indices(IA, I)
+
+checks whether the "requested" indices in the tuple `I` fall within
+the bounds of the "permitted" indices specified by the tuple
+`IA`. This function iteratively consumes elements of these tuples,
+usually in a 1-for-1 fashion,
+
+    checkbounds_indices((IA1, IA...), (I1, I...)) = checkindex(Bool, IA1, I1) &
+                                                    checkbounds_indices(IA, I)
 
-Return `true` if the specified `indexes` are in bounds for the given `array`. Subtypes of
-`AbstractArray` should specialize this method if they need to provide custom bounds checking
-behaviors.
+Note that `checkindex` is being used to perform the actual
+bounds-check for a single dimension of the array.
+
+There are two import exceptions to the 1-1 rule: linear indexing and
+CartesianIndex{N}, both of which may "consume" more than one element
+of `IA`.
 """
```

Comment:
the docstring should be attached to the most general signature (or a zero-method line)